### ISSUE TITLE
Pixel-perfect mode does not respect render.aspect setting (was: Incorrect aspect ratio in Waxworks)

### ISSUE THREAD

**Issue by veksha** [dosbox-staging/dosbox-staging#544]
**To Reproduce**
config:
    fullscreen=true
    fullresolution=1920x1080
    output=texturepp
launch Waxworks game

**Expected behaviour**
aspect ratio in fullscreen mode should match aspect ratio in window mode (press alt-enter to compare)
BTW, fullresolution=1024x768 gives correct aspect ratio.

**Environment (please complete the following information):**
- Windows 10
- AMD Ryzen 7 2700

**Version of dosbox-staging:**
0.75.0

**Additional context**
Add any other context or details about the problem you might found relevant
(e.g. screenshots, other dosbox versions impacted, etc).
![waxworks1](https://user-images.githubusercontent.com/275333/90306047-dd446680-ded1-11ea-9c44-eac8a1e3ad73.png)
![waxworks2](https://user-images.githubusercontent.com/275333/90306046-dc133980-ded1-11ea-83ba-bf581b6cfd70.png)

```
dosbox-staging version 0.75.0
---
CONFIG: Loading primary file C:\Users\yura\AppData\Local\OGDBL\dosbox.conf
CONFIG: Unknown option fulldouble
CONFIG: Unknown option autolock
CONFIG: Unknown option usescancodes
CONFIG: Loading additional file D:\Games\_DOS\waxworks\wax.dosbox
MAIN: Draw resolution: 640x480, pixel aspect ratio: 1.00
MAIN: Emulator resolution: 1920x1080
MAIN: Pixel-perfect scaling (2x2): 640x480 (PAR 1.00) -> 1280x960 (PAR 1.00)
SDL: Using driver "direct3d" for texture renderer
SDL: Mouse will move seamlessly without being captured.
MIXER: Got different values from SDL: freq 44100, blocksize 1024
MIDI: Opened device:win32
MAPPER: Found no joysticks
MAPPER: Loaded default key bindings
SDL: captured the mouse
MAIN: Draw resolution: 640x400, pixel aspect ratio: 1.20
MAIN: Emulator resolution: 1920x1080
MAIN: Pixel-perfect scaling (2x2): 640x400 (PAR 1.20) -> 1280x800 (PAR 1.00)
SDL: Using driver "direct3d" for texture renderer
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
MAIN: Emulator resolution: 1920x1080
MAIN: Pixel-perfect scaling (4x5): 320x200 (PAR 1.20) -> 1280x1000 (PAR 1.25)
SDL: Using driver "direct3d" for texture renderer
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
MAIN: Emulator resolution: 320x240
MAIN: Pixel-perfect scaling (1x1): 320x200 (PAR 1.20) -> 320x200 (PAR 1.00)
SDL: Using driver "direct3d" for texture renderer
SDL: released the mouse
```

**Comment by dreamer** [comment]
> aspect ratio in fullscreen mode should match aspect ratio in window mode (press alt-enter to compare)

No, it should not - because you selected `output=texturepp` - this explicitly asks dosbox-staging to forget about preserving aspect ratio to pursue perfect integer scaling.

Try using:
```
[sdl]
fullscreen=true
fullresolution=desktop
output=opengl

[render]
glshader=sharp
```
and please respond if these settings give you the effect you expected :)

(these are all default settings in dosbox-staging btw)

**Comment by veksha** [comment]
> > aspect ratio in fullscreen mode should match aspect ratio in window mode (press alt-enter to compare)
> No, it should not - because you selected `output=texturepp` - this explicitly asks dosbox-staging to forget about preserving aspect ratio to pursue perfect integer scaling.

I noticed that it works as expected for emulator resolution 1280x720 and lower (1024x768, 800x600)

```
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
MAIN: Emulator resolution: 1280x720
MAIN: Pixel-perfect scaling (3x3): 320x200 (PAR 1.20) -> 960x600 (PAR 1.00)

```

but for 1280x1024, 1920x1080 it is scaling like 4x5 and not like 3x3.

```
MAIN: Draw resolution: 320x200, double-width, double-height, pixel aspect ratio: 1.20
MAIN: Emulator resolution: 1920x1080
MAIN: Pixel-perfect scaling (4x5): 320x200 (PAR 1.20) -> 1280x1000 (PAR 1.25)
```

so aspect ratio is preserved on low resolutions somehow.
i need my fullhd resolution to work because it is native for my desktop (alt-tabbing is quick and all my windows will not resize like crazy after dosbox shutdown)


> 
> Try using:
> 
> ```
> [sdl]
> fullscreen=true
> fullresolution=desktop
> output=opengl
> 
> [render]
> glshader=sharp
> ```
> and please respond if these settings give you the effect you expected :)

this is working, but this is not integer scaling.

**Comment by dreamer** [comment]
> but for 1280x1024, 1920x1080 it is scaling like 4x5 and not like 3x3.

That's **by design** because the vast majority of DOS games using 320x200 resolution were designed for 4:3 screens with 1.20 Pixel Aspect Ratio. Thus the algorithm picks 1.25 PAR, as for *those games* it would provide less distortion than 1.00.

However, after a closer look at the screenshots you provided, it seems like this particular game belongs to the minority, that was designed for 1.00 PAR (square pixels). Use the following settings to force square pixels:

```
[render]
aspect = false
```
I hope this gives exactly the effect you want and game was designed for :)

I would love to see other screenshots as well - can you look for other square or round objects in the game?

**Comment by veksha** [comment]
> > but for 1280x1024, 1920x1080 it is scaling like 4x5 and not like 3x3.
> [render]
> aspect = false
> ```
> I hope this gives exactly the effect you want and game was designed for :)

i have already played with this switch and unfortunately it changes nothing.
it would be nice to have this option working for this game. maybe it works for other games. i don't know.

> I would love to see other screenshots as well - can you look for other square or round objects in the game?

i found few screenshots on scummvm website. here we see distortion in aspect ratio (640x480)
https://www.scummvm.org/screenshots/agos/waxworks?lang=en

but all other screenshots on google image search page is perfect 1:1 ratio (640x400)

**Comment by dreamer** [comment]
OK, I tested the game and well… I was right the first time - the game was designed for 4:3 screen and pixel aspect ratio 1.2, therefore the dosbox-staging configured to use `texturepp` **correctly** picked up PAR 1.25 to scale pixels 4x5 when the screen is large enough to accommodate it.

Here are screens from the very beginning of the game to prove it:

Using 4:3 aspect ratio with non-square pixels (PAR 1.2)
![Screenshot from 2020-08-15 15-28-47](https://user-images.githubusercontent.com/3967/90313652-92950f80-df0e-11ea-925a-8016027cb7ba.png)

Using 16:10 aspect ratio with square pixels (`aspect=true`)
![Screenshot from 2020-08-15 15-40-44](https://user-images.githubusercontent.com/3967/90313661-9aed4a80-df0e-11ea-853f-d08826d36178.png)

Orb is clearly stretched horizontally in the second screenshot.

That means:

- dosbox-staging with default settings works correctly
-  ScummVM is correct to stretch the game to 4:3 by default, as seen on ScummVM site
- dosbox-staging configured with `texturepp` works correctly as well and picks the PAR appropriate to the available screen space
- all the screenshots showing the game in 320x200 or 640x400 with square pixels are **incorrect**

**However**, I can confirm that when `aspect=false` and `output=texturepp` are both selected, algorithm should *always* prefer square pixels but it does not, therefore we have a bug.

@veksha Thanks for testing; I have a feeling this should be a quick fix :) I'll eddit the issue title to reflect the real bug.

**Comment by kcgen** [comment]
When `aspect=true` is set, that's equivalent to saying "_multiply the game's vertical image by 1.2_". 

All the `*pp` modes respect that request, but need to be given a large enough screen canvas to achieve it.  

For example, if the game generates a 320x200 picture w/ aspect=true, then the result is 320x (200 * 1.2) = 320 x240; but those 240 pixels arent whole integers anymore and will contain blending or distortion. 

The pp modes need to turn that 1.2 multiplication  into a whole number, so the 1.2 is converted to 5, and the width is scaled by 4.  5/4 = 1.2

(320 * 4) x (200 * 5) = 1280 x 1000

So the pp modes need to be given that much drawing realestate (via windowresolution or via fullscreen) to show it. 

1280x1000 is pretty close to vertical limits of a 1080p screen... So games that generate even bigger native images (320x240, or 640x400) simply won't work with pp unless you have a massive screen.  

The `*nb` modes are more fogiving by allow for some leeway in the calculation multipliers, so they are generally good choices for those with normal 1080p screens but who want sharp pixels even only non 320x240 games

**Comment by dreamer** [comment]
@kcgen It seems like our pixel-perfect implementation respects `aspect=true`, but does not respect `aspect=false`.

**Comment by veksha** [comment]
I still think human bodies are too tall and stretched, but it can be just me. :-)

thank you for clarification. I increased windowresolution and now it matches aspect ratio of fullscreen mode. didn't know about this "silent fallback" of drawing without aspect ratio when there is not enough place for all the pixels. i expected that pixels just will go beyond the screen, but maybe this design is even better to do this fallback so no pixels are lost.

and i'm glad we found something was not working.

**Comment by dreamer** [comment]
@veksha Here's a build to test: https://github.com/dosbox-staging/dosbox-staging/actions/runs/209904117. Try using settings:
```
[sdl]
fullscreen=true
fullresolution=desktop
output=texturepp

[render]
aspect=false
```
I hope it will have perfectly square pixels regardless of your screen resolution :)

For resolution 1920x1080 it should result in 5x5 pixels with a black border around the game screen.

**Comment by veksha** [comment]
it works. thank you!

**Comment by dreamer** [comment]
Great :) I think this fix is small enough to squeeze into 0.75.1 release. I'll prepare PR tomorrow.

**Comment by dreamer** [comment]
The fix will be included in upcoming 0.75.1 release.

@veksha Thanks for report and testing :)

**Comment by kcgen** [comment]
> @kcgen It seems like our pixel-perfect implementation respects aspect=true, but does not respect aspect=false.

Fixed in [prior](https://github.com/dosbox-staging/dosbox-staging/pull/1491), PP now respects `aspect=false` (or `true`):

![Screenshot from 2022-04-28 14-13-28](https://user-images.githubusercontent.com/1557255/165847207-4cd553c4-51ac-4961-82e2-b9e733cdeec1.png)
![Screenshot from 2022-04-28 14-13-56](https://user-images.githubusercontent.com/1557255/165847210-0f8bec9e-5439-4607-834f-e5a43810a873.png)
